File: C++/Day04.cpp
#ifdef DAY04

#include "AH.h"

namespace Day04
{
	typedef std::vector<std::string> Grid;
	
	int findWord(const Grid & g, const std::string target)
	{
		std::string revTarget = target;
		std::reverse(revTarget.begin(), revTarget.end());

		const size_t rows = g.size();
		const size_t cols = g[0].size();
		const size_t target_size = target.size();

		int counter = 0;
		// W->E & E->W
		for (size_t r = 0; r < rows; r++) {
			for (size_t c = 0; c < cols - target_size + 1; c++) {
				bool good = true;
				for (size_t idx = 0; idx < target_size; idx++) {
					if (g[r][c + idx] != target.at(idx)) {
						good = false;
						break;
					}
				}
				if (good) {
					counter++;
				}
			
				good = true;
				for (size_t idx = 0; idx < target_size; idx++) {
					if (g[r][c + idx] != revTarget.at(idx)) {
						good = false;
						break;
					}
				}
				if (good) {
					counter++;
				}
			 }
		}

		//N->S & S->N
		for (size_t r = 0; r < rows - target_size + 1; r++) {
			for (size_t c = 0; c < cols; c++) {
				bool good = true;
				for (size_t idx = 0; idx < target_size; idx++) {
					if (g[r + idx][c] != target.at(idx)) {
						good = false;
						break;
					}
				}
				if (good) {
					counter++;
				}
			
				good = true;
				for (size_t idx = 0; idx < target_size; idx++) {
					if (g[r + idx][c] != revTarget.at(idx)) {
						good = false;
						break;
					}
				}
				if (good) {
					counter++;
				}
			 }
		}

		//NW->SE & SE->NW
		for (size_t r = 0; r < rows - target_size + 1; r++) {
			for (size_t c = 0; c < cols - target_size + 1; c++) {
				bool good = true;
				for (size_t idx = 0; idx < target_size; idx++) {
					if (g[r + idx][c + idx] != target.at(idx)) {
						good =false;
						break;
					}
				}
				if (good) {
					counter++;
				}
				
				good = true;
				for (size_t idx = 0; idx < target_size; idx++) {
					if (g[r + idx][c + idx] != revTarget.at(idx)) {
						good = false;
						break;
					}
				}
				if (good) {
					counter++;
				}
			}
		}

		//NE->SW & SW->NE
		for (size_t r = 0; r < rows - target_size + 1; r++) {
			for (size_t c = target_size - 1; c < cols; c++) {
				bool good = true;
				for (size_t idx = 0; idx < target_size; idx++) {
					if (g[r + idx][c - idx] != target.at(idx)) {
						good =false;
						break;
					}
				}
				if (good) {
					counter++;
				}
				
				good = true;
				for (size_t idx = 0; idx < target_size; idx++) {
					if (g[r + idx][c - idx] != revTarget.at(idx)) {
						good = false;
						break;
					}
				}
				if (good) {
					counter++;
				}
			}
		}

		return counter;
	}

	int findX(const Grid & g)
	{
		const size_t rows = g.size();
		const size_t cols = g[0].size();
		int counter = 0;

		for (size_t r = 1; r < rows - 1; r++) {
			for (size_t c = 1; c < cols - 1; c++) {
				if (g[r][c] != 'A') { continue; }

				if (
					(((g[r-1][c-1] == 'M') && (g[r+1][c+1] == 'S')) ||
					 ((g[r-1][c-1] == 'S') && (g[r+1][c+1] == 'M'))) &&
					(((g[r-1][c+1] == 'M') && (g[r+1][c-1] == 'S')) ||
					 ((g[r-1][c+1] == 'S') && (g[r+1][c-1] == 'M')))
				) {
					counter++;
				}
			}
		}

		return counter;
	}

	int Run(const std::string& filename)
	{
		auto g = AH::ReadTextFile(filename);
		auto p1 = findWord(g, "XMAS");
		auto p2 = findX(g);

		AH::PrintSoln(4, p1, p2);

		return 0;	
	}

}

#endif

File: Go/Day04/Day04.go
//go:build d04
// +build d04

package Day04

import (
	AH "AoC2024/adventhelper"
)

type Offsets struct {
	row_s int
	row_e int
	col_s int
	col_e int
	row_i int
	col_i int
}

func findWord2(grid []string, target string) (counter int) {
	revTarget := AH.ReverseString(target)
	rows := len(grid)
	cols := len(grid[0])
	lent := len(target)

	counter = 0

	offs := []Offsets{
		Offsets{0, rows, 0, cols - lent + 1, 0, 1},
		Offsets{0, rows - lent + 1, 0, cols, 1, 0},
		Offsets{0, rows - lent + 1, 0, cols - lent + 1, 1, 1},
		Offsets{0, rows - lent + 1, lent - 1, cols, 1, -1},
	}

	for _, off := range offs {
		for r := off.row_s; r < off.row_e; r++ {
			for c := off.col_s; c < off.col_e; c++ {
				ok_pos := true
				ok_neg := true
				for idx := 0; idx < lent; idx++ {
					rn := grid[r+off.row_i*idx][c+off.col_i*idx]
					ok_pos = ok_pos && (rn == target[idx])
					ok_neg = ok_neg && (rn == revTarget[idx])

					if !(ok_neg || ok_pos) {
						break
					}
				}
				if ok_pos {
					counter++
				}
				if ok_neg {
					counter++
				}
			}
		}
	}

	return
}

func findX(grid []string) (counter int) {
	rows := len(grid)
	cols := len(grid[0])
	counter = 0

	for r := 1; r < rows-1; r++ {
		for c := 1; c < cols-1; c++ {
			if grid[r][c] != 'A' {
				continue
			}

			if (((grid[r-1][c-1] == 'M') && (grid[r+1][c+1] == 'S')) ||
				((grid[r-1][c-1] == 'S') && (grid[r+1][c+1] == 'M'))) &&
				(((grid[r-1][c+1] == 'M') && (grid[r+1][c-1] == 'S')) ||
					((grid[r-1][c+1] == 'S') && (grid[r+1][c-1] == 'M'))) {
				counter++
			}

		}
	}

	return
}

func Run() {
	g, _ := AH.ReadStrFile("../input/input04.txt")

	part1, part2 := findWord2(g, "XMAS"), findX(g)

	AH.PrintSoln(1, part1, part2)

	return
}
